libnetwork/iptables/firewalld.go

```diff
@@ -52,11 +58,43 @@ func firewalldInit() error {
 	if connection, err = newConnection(); err != nil {
 		return fmt.Errorf("Failed to connect to D-Bus system bus: %v", err)
 	}
-	firewalldRunning = checkRunning()
+
+	firewalldRunning, err = checkRunning()
+
+	if err != nil {
+		return err
+	}
+	// If firewalld is not running on the first check poll
 	if !firewalldRunning {
-		connection.sysconn.Close()
-		connection = nil
+		tick := time.NewTicker(500 * time.Millisecond)
+		timeLapse := time.Duration(0)
+		defer tick.Stop()
+
+	TickLoop:
+		for {
+			select {
+			case <-tick.C:
```

Comment:
Still slightly wondering if this loop would've made sense to be part of `checkRunning` as the "actual" check if it's running is in the goroutine inside that function, so perhaps retrying it as part of that function would work? 

w.r.t. "dbus connections  hanging";  that function could set a global context-timeout of 2.5 seconds (or whatever we want the maximum to be), and create a per-invocation context from that (how long we want each check to take as maximum).

Reply:
I change the implementation to use connection.sysObj.CallWithContext ven though I observed it hang.  It should work for the time being.

Reply:
Are you able to reproduce the "hang"? Does that mean the command never returns with an expected value (i.e., we must do the retry loop)?

Given that we have a context available now, I'm wondering if we'd be able to create a separate context-timeout for each invocation, and if canceling the context cancels that invocation, we could use that to time-constraint "bad" calls (if such calls never return a useful result); it would be something like (very invalid code, but just to roughly describe my thinking);

```go
// allow all of it to take X seconds
ctx, cancel := context.WithTimeout(ctx, 10 * time.Second)
defer cancel()

for .... {
    status, err := func(ctx context.Context) (result, error){
        // limit invocation to 1 second max
        ctx, cancel := context.WithTimeout(ctx, 1 * time.Second)
        defer cancel()
        err = connection.sysObj.CallWithContext(ctx, ....)
        return ....
    }(ctx)
    if status == succesful {
        break
    }

}
```

Reply:
I am not able to reproduce the problem consistently so I'm not certain its worth making a modification until there is a consistently reproducible result.  I was able to produce the problem in a number of ways but not repeatedly the same way.   I can only reproduce one version of the "hanging" problem consistently and repeatedly.

1.) starting a system with a failed polkit and dbus
2.) starting firewalld successfully

The other scenarios involved different failures between policy kit, dbus and firewalld but don't happen consistently enough to warrant changes unless someone has automation to determine how often the problem occurs. Is is 1/10, 1/100, 1/1000 times. I don't know and I don't know if we have the time to investigate that problem. I don't have the automation and I can't prove it with any certainty.

Reply:
One other note, I believe the existing code allows for making checkRunning() calls at regular time intervals. The code you have doesn't make calls at regular time intervals (as a ticker allows), it only limits:

1.) The amount of time all calls can take in aggregate
2.) The amount of time individual calls can last


I don't see how it solves the issue of dockerd hanging with no result being returned unless we know more about what caueses dockerd to hang other than the one issue I can reproduce and there is an existing solution for.

Reply:
> As far as I could tell **connection.sysObj.CallWithContext** does not return a value at all, under the right circumstances

What does "not return[ing] a value at all" mean here? `CallWithContext` is a Go function with a result of type `*dbus.Call`, so it has to return some result value when it returns. Did you mean:
1. `connection.sysObj.CallWithContext` _blocks forever_, i.e. does not return, period,
2. `connection.sysObj.CallWithContext` returns `(*dbus.Call)(nil)`, or
3. `connection.sysObj.CallWithContext` returns a `*dbus.Call` containing an error response, resulting in the chained `.Store(&currentState)` call not modifying the `currentState` variable?

Reply:
@corhere I cannot reproduce the issue without a massive effort but I assume he meant blocks forever, does not return period.  When I tested this it blocked for thirty minutes.  I occassionally saw this behavior in the original checkRunning implementation as well but it was extremely rare.

Reply:
@psaintlaurent Why are you referring to yourself in the second person? [You wrote the comment](https://github.com/moby/moby/pull/48043#discussion_r1702137001) I quoted and am asking for clarification on.

Does the call continue to block forever even after the context becomes done? If so, that's an egregious bug in godbus. Otherwise, @thaJeztah's suggestion of setting a timeout on each call to limit how long it can hang for for sounds like an ideal solution to the hanging problem.

Reply:
@corhere I apologize if it came across as if I was referring to myself in the second person.  I did not make the original comment regarding "not returning a value". I never observed anything other than a hanging of dockerd so I used the wording "hang" until I was asked about whether the value was returned.